pymc/stats/model_comparison.py

```diff
@@ -0,0 +1,126 @@
+    var_names: sequence of str, optional
+        List of Observed variable names for which to compute log_likelihood. Defaults to all observed variables
+    extend_inferencedata: bool, default True
+        Whether to extend the original InferenceData or return a new one
+    model: Model, optional
+    sample_dims: sequence of str, default ("chain", "draw")
+    progressbar: bool, default True
+
+    Returns
+    -------
+    idata: InferenceData
+        InferenceData with log_likelihood group
+
+    """
+
+    posterior = idata["posterior"]
+
+    model = modelcontext(model)
+
+    if var_names is None:
+        observed_vars = model.observed_RVs
+        var_names = tuple(rv.name for rv in observed_vars)
+    else:
+        observed_vars = [model.named_vars[name] for name in var_names]
+        if not set(observed_vars).issubset(model.observed_RVs):
+            raise ValueError(f"var_names must refer to observed_RVs in the model. Got: {var_names}")
+
+    # We need to temporarily disable transforms, because the InferenceData only keeps the untransformed values
+    # pylint: disable=used-before-assignment
+    try:
+        original_rvs_to_values = model.rvs_to_values
+        original_rvs_to_transforms = model.rvs_to_transforms
+
+        model.rvs_to_values = {
+            rv: rv.clone() if rv not in model.observed_RVs else value
+            for rv, value in model.rvs_to_values.items()
+        }
+        model.rvs_to_transforms = {rv: None for rv in model.basic_RVs}
+
+        elemwise_loglike_fn = model.compile_fn(
+            inputs=model.value_vars,
+            outs=model.logp(vars=observed_vars, sum=False),
+            on_unused_input="ignore",
+        )
+    finally:
+        model.rvs_to_values = original_rvs_to_values
+        model.rvs_to_transforms = original_rvs_to_transforms
+    # pylint: enable=used-before-assignment
+
+    # Ignore Deterministics
+    posterior_values = posterior[[rv.name for rv in model.free_RVs]]
+    posterior_pts, stacked_dims = dataset_to_point_list(posterior_values, sample_dims)
+    n_pts = len(posterior_pts)
+    loglike_dict = _DefaultTrace(n_pts)
+    indices = range(n_pts)
+    if progressbar:
+        indices = progress_bar(indices, total=n_pts, display=progressbar)
+
+    for idx in indices:
+        loglikes_pts = elemwise_loglike_fn(posterior_pts[idx])
+        for rv_name, rv_loglike in zip(var_names, loglikes_pts):
+            loglike_dict.insert(rv_name, rv_loglike, idx)
+
+    loglike_trace = loglike_dict.trace_dict
+    for key, array in loglike_trace.items():
+        loglike_trace[key] = array.reshape(
+            (*[len(coord) for coord in stacked_dims.values()], *array.shape[1:])
+        )
+
+    idata_loglike = InferenceData(
+        log_likelihood=dict_to_dataset(
+            loglike_trace,
+            library=pymc,
+            dims=posterior.dims,
+            coords=posterior.coords,
+            default_dims=list(sample_dims),
+        )
+    )
```

Comment:
Main most important comment: I think this won't preserve the coordinates and dimensions which would be an undesired backward change. I am not even sure it will work, `posterior.dims` is not a dict with variable names as keys and dimensions of those variables as values, and if it were it would not have the variable names in log_likelihood.

---

Extra comment, not sure if it might be helpful. It would be very similar to do:

```suggestion
    log_likelihood=dict_to_dataset(
        loglike_trace,
        library=pymc,
        dims=posterior.dims,
        coords=posterior.coords,
        default_dims=list(sample_dims),
    )
```

Then return the dataset (very close to returning a dict with the pointwise log likelihood values) if extend=False, otherwise use [add_groups](https://python.arviz.org/en/stable/api/generated/arviz.InferenceData.add_groups.html). I am not sure what are the use cases envisioned with `extend=False` so I don't really see pros and cons of each option.

Reply:
Sounds cleaner. I'll check about the dims

Reply:
Fixed